common/key_value_database/platform/lmdb/lmdb.cpp

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <iostream>
+#include <lmdb++.h>
+#include <cstdio>
+
+#include <common/common.hpp>
+#include <common/key_value_database_lmdb.hpp>
+
+namespace mender::common::key_value_database {
+
+class LmdbTransaction : public Transaction {
+public:
+	LmdbTransaction(lmdb::txn &txn, lmdb::dbi &dbi);
+
+	expected::ExpectedBytes Read(const string &key) override;
+	error::Error Write(const string &key, const vector<uint8_t> &value) override;
+	error::Error Remove(const string &key) override;
+
+private:
+	lmdb::txn &txn_;
+	lmdb::dbi &dbi_;
+};
+
+LmdbTransaction::LmdbTransaction(lmdb::txn &txn, lmdb::dbi &dbi) :
+	txn_(txn),
+	dbi_(dbi) {
+}
+
+expected::ExpectedBytes LmdbTransaction::Read(const string &key) {
+	try {
+		std::string_view value;
+		bool exists = dbi_.get(txn_, key, value);
+		if (!exists) {
+			return MakeError(KeyError, "Key " + key + " not found in database");
+		}
+
+		return common::ByteVectorFromString(value);
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+}
+
+error::Error LmdbTransaction::Write(const string &key, const vector<uint8_t> &value) {
+	try {
+		string_view data(reinterpret_cast<const char *>(value.data()), value.size());
+		bool exists = dbi_.put(txn_, key, data);
+		if (!exists) {
+			return MakeError(AlreadyExistsError, "Key " + key + " already exists");
+		}
+
+		return error::NoError;
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+}
+
+error::Error LmdbTransaction::Remove(const string &key) {
+	try {
+		// We don't treat !exists as an error, just ignore return code.
```

Comment:
Sounds like a debug log?

Reply:
In this case I wouldn't. Removing something non-existing is completely normal, IMHO. If it matters, then the caller should call `Read` first to check instead.